File: app/src/main/java/algorithm/study/devjk_sample/leetcode/Let_229.java
package algorithm.study.devjk_sample.leetcode;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

/**
 * 리트코드 - Majority Element II
 *
 * @link https://leetcode.com/problems/majority-element-ii
 */
public class Let_229 {

  public static void main(String[] args){
    Let_229 sol = new Let_229();
    System.out.println(sol.majorityElement(new int[]{3, 2, 3}));
    System.out.println(sol.majorityElement(new int[]{1}));
    System.out.println(sol.majorityElement(new int[]{1, 2}));
    System.out.println(sol.majorityElement(new int[]{2, 2, 1, 3}));
  }

  public List<Integer> majorityElement(int[] nums) {
    List<Integer> list = new ArrayList<>();
    nums = Arrays.stream(nums).sorted().toArray();
    int target = nums.length / 3;
    int cnt = 0;
    int before = nums[0];
    for(int num : nums) {
      if(num == before) {
        cnt++;
      }else {
        if(cnt > target) {
          list.add(before);
        }
        cnt = 1;
      }
      before = num;
    }
    if(cnt > target) {
      list.add(before);
    }
    return list;
  }

}

File: app/src/main/java/algorithm/study/devjk_sample/leetcode/Let_229.kt
package algorithm.study.devjk_sample.leetcode

/**
 * 리트코드 - Majority Element II
 *
 * @link https://leetcode.com/problems/majority-element-ii
 */
class Let_229A {

    fun majorityElement(nums: IntArray): List<Int> {
        val ans : MutableList<Int> = mutableListOf()
        val len = nums.size / 3
        nums.sort()

        var before = 1000000001
        var i = 0
        var cnt = 0
        while(i < nums.size) {
            val num = nums[i]
            if(before != num) {
                before = num
                cnt++
                i++
                while(i < nums.size && nums[i] == before) {
                    cnt++
                    i++
                }
                if(cnt > len) {
                    ans.add(before)
                }
                cnt = 0
            }
        }
        return ans
    }

}

fun main(args: Array<String>) {
    val sol = Let_229A()
    println(sol.majorityElement(intArrayOf(3, 2, 3)))
    println(sol.majorityElement(intArrayOf(1)))
    println(sol.majorityElement(intArrayOf(1, 2)))
    println(sol.majorityElement(intArrayOf(1, 2, 3)))
}